File: android/app/src/main/kotlin/com/pixeldev/Docly/pdf/AnnotationHandler.kt
package com.pixeldev.Docly.pdf

data class HighlightOperation(
    val id: String,
    val page: Int,
    val x: Double,
    val y: Double,
    val w: Double,
    val h: Double,
    val color: Int,
    val opacity: Double
)

data class TextOperation(
    val id: String,
    val page: Int,
    val text: String,
    val x: Double,
    val y: Double,
    val color: Int = 0xFF000000.toInt(),
    val fontSize: Double = 16.0
)

data class ImageOperation(
    val id: String,
    val page: Int,
    val path: String,
    val x: Double,
    val y: Double,
    val width: Double = 100.0,
    val height: Double = 100.0
)

class AnnotationHandler {
    private val highlights = mutableListOf<HighlightOperation>()
    private val texts = mutableListOf<TextOperation>()
    private val images = mutableListOf<ImageOperation>()

    fun addHighlight(op: HighlightOperation) {
        highlights += op
    }

    fun updateHighlight(id: String, op: HighlightOperation) {
        val index = highlights.indexOfFirst { it.id == id }
        if (index >= 0) {
            highlights[index] = op
        }
    }

    fun deleteHighlight(id: String) {
        highlights.removeAll { it.id == id }
    }

    fun addText(op: TextOperation) {
        texts += op
    }

    fun updateText(id: String, op: TextOperation) {
        val index = texts.indexOfFirst { it.id == id }
        if (index >= 0) {
            texts[index] = op
        }
    }

    fun deleteText(id: String) {
        texts.removeAll { it.id == id }
    }

    fun addImage(op: ImageOperation) {
        images += op
    }

    fun updateImage(id: String, op: ImageOperation) {
        val index = images.indexOfFirst { it.id == id }
        if (index >= 0) {
            images[index] = op
        }
    }

    fun deleteImage(id: String) {
        images.removeAll { it.id == id }
    }

    fun highlightOperations(): List<HighlightOperation> = highlights.toList()

    fun textOperations(): List<TextOperation> = texts.toList()

    fun imageOperations(): List<ImageOperation> = images.toList()

    fun clear() {
        highlights.clear()
        texts.clear()
        images.clear()
    }
}

File: ios/Runner/pdf/AnnotationHandler.swift
import Foundation

struct HighlightOperation {
  let id: String
  let page: Int
  let x: Double
  let y: Double
  let w: Double
  let h: Double
  let color: Int
  let opacity: Double
}

struct TextOperation {
  let id: String
  let page: Int
  let text: String
  let x: Double
  let y: Double
  let color: Int
  let fontSize: Double
}

struct ImageOperation {
  let id: String
  let page: Int
  let path: String
  let x: Double
  let y: Double
  let width: Double
  let height: Double
}

final class AnnotationHandler {
  private var highlights: [HighlightOperation] = []
  private var texts: [TextOperation] = []
  private var images: [ImageOperation] = []

  func addHighlight(_ op: HighlightOperation) {
    highlights.append(op)
  }

  func updateHighlight(id: String, op: HighlightOperation) {
    if let index = highlights.firstIndex(where: { $0.id == id }) {
      highlights[index] = op
    }
  }

  func deleteHighlight(id: String) {
    highlights.removeAll { $0.id == id }
  }

  func addText(_ op: TextOperation) {
    texts.append(op)
  }

  func updateText(id: String, op: TextOperation) {
    if let index = texts.firstIndex(where: { $0.id == id }) {
      texts[index] = op
    }
  }

  func deleteText(id: String) {
    texts.removeAll { $0.id == id }
  }

  func addImage(_ op: ImageOperation) {
    images.append(op)
  }

  func updateImage(id: String, op: ImageOperation) {
    if let index = images.firstIndex(where: { $0.id == id }) {
      images[index] = op
    }
  }

  func deleteImage(id: String) {
    images.removeAll { $0.id == id }
  }

  func allHighlights() -> [HighlightOperation] {
    highlights
  }

  func allTexts() -> [TextOperation] {
    texts
  }

  func allImages() -> [ImageOperation] {
    images
  }

  func clear() {
    highlights.removeAll()
    texts.removeAll()
    images.removeAll()
  }
}
